File: college/sem4/ICS225-DATASTRUCTURES-III/lab/splay/splaytrees.py
from enum import Enum
from copy import copy
from typing import List


class Movement(Enum):
    LEFT = 0
    RIGHT = 1

class Node:
    def __init__(self, val: int):
        self.val: int = val
        self.left: Node | None = None
        self.right: Node | None = None

    def insert(self, val):
        if val < self.val:
            if self.left:
                self.left.insert(val)
            else:
                self.left = Node(val)
        else:
            if self.right:
                self.right.insert(val)
            else:
                self.right = Node(val)

    def inorder(self):
        if self.left:
            self.left.inorder()
        print(self.val, end=" ")
        if self.right:
            self.right.inorder()

class Splay:
    @staticmethod
    def zig(p: Node | None, x: Node | None) -> Node | None:
        if not p or not x: return
        p.left = x.right
        x.right = p
        return x

    @staticmethod
    def zag(p: Node | None, x: Node | None) -> Node | None:
        if not p or not x: return
        p.right = x.left
        x.left = p
        return x

    @staticmethod
    def zigzig(g: Node | None, p: Node | None, x: Node | None) -> Node | None:
        return Splay.zig(Splay.zig(g, p), x)

    @staticmethod
    def zagzag(g: Node | None, p: Node | None, x: Node | None) -> Node | None:
        return Splay.zag(Splay.zag(g, p), x)

    @staticmethod
    def zigzag(g: Node | None, p: Node | None, x: Node | None) -> Node | None:
        if not g or not p or not x: return
        g.left = Splay.zag(p, x)
        return Splay.zig(g, x)

    @staticmethod
    def zagzig(g: Node | None, p: Node | None, x: Node | None) -> Node | None:
        if not g or not p or not x: return
        g.right = Splay.zig(p, x)
        return Splay.zag(g, x)

    @staticmethod
    def insert(root: Node | None, val) -> Node | None:
        if not root: return
        root.insert(val)

        root = Splay.find(root, val)
        return root

    @staticmethod
    def split(root: Node | None, k: int) -> None | List[Node | None]:
        root = Splay.find(root, k)
        if not root: return

        return [root.left, root.right]

    @staticmethod
    def join(left: Node | None, right: Node | None) -> Node | None:
        if not left:
            return right

        max = copy(left)
        while max.right:
            max = max.right

        left = Splay.find(left, max.val)
        if not left: return

        left.right = right
        return left

    @staticmethod
    def delete(root: Node | None, val) -> Node | None:
        if not root: return

        root = Splay.find(root, val)
        if not root: return
        root = Splay.join(root.left, root.right)

        return root

    @staticmethod
    def find(root: Node | None, val: int, i = 0) -> Node | None:
        i += 1
        if not root: return
        d1: Movement = Movement.LEFT
        d2: Movement = Movement.LEFT
        if root.val == val:
            return root
        
        g = root
        if g.val < val:
            p = g.right
            d1 = Movement.RIGHT
        else:
            p = g.left
            d1 = Movement.LEFT
        if not p: return


        if p.val == val:
            if d1 == Movement.LEFT:
                g = Splay.zig(g, p)
                return g
            else:
                g = Splay.zag(g, p)
                return g

        if p.val < val:
            x = p.right
            d2 = Movement.RIGHT
        else:
            x = p.left
            d2 = Movement.LEFT
        if not x: return

        if p.val != val and x.val != val:
            x = Splay.find(x, val, i)
            if not x: return

        if x.val == val:
            if d1 == Movement.LEFT:
                if d2 == Movement.LEFT:
                    g = Splay.zigzig(g, p, x)
                    return g
                else:
                    g = Splay.zigzag(g, p, x)
                    return g
            else:
                if d2 == Movement.LEFT:
                    g = Splay.zagzig(g, p, x)
                    return g
                else:
                    g = Splay.zagzag(g, p, x)
                    return g

tree = Node(22)
vals = [10, 30, 6, 18, 27, 9, 15, 19, 12, 14]
for val in vals:
    tree.insert(val)

tree.inorder()
print()

# Put these here only
# tree = Splay.delete(tree, 22) -> To delete
# tree = Splay.find(tree, 22) -> To bring it to top
# tree = Splay.insert(tree, 69) -> To Insert and bring it to top

if tree: tree.inorder()
print()

if tree: print(tree.val)

File: college/sem4/ICS225-DATASTRUCTURES-III/lab/splay/splaytrees.cpp
#include <iostream>
#include <vector>

enum class Movement { LEFT, RIGHT };

class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node(int val) : val(val), left(nullptr), right(nullptr) {}

    ~Node() {
        delete left;
        delete right;
    }

    void insert(int val) {
        if (val < this->val) {
            if (left) {
                left->insert(val);
            } else {
                left = new Node(val);
            }
        } else {
            if (right) {
                right->insert(val);
            } else {
                right = new Node(val);
            }
        }
    }

    void inorder() {
        if (left) {
            left->inorder();
        }
        std::cout << val << " ";
        if (right) {
            right->inorder();
        }
    }
};

class Splay {
public:
    static Node* zig(Node* p, Node* x) {
        if (!p || !x) return nullptr;
        p->left = x->right;
        x->right = p;
        return x;
    }

    static Node* zag(Node* p, Node* x) {
        if (!p || !x) return nullptr;
        p->right = x->left;
        x->left = p;
        return x;
    }

    static Node* zigzig(Node* g, Node* p, Node* x) {
        return zig(zig(g, p), x);
    }

    static Node* zagzag(Node* g, Node* p, Node* x) {
        return zag(zag(g, p), x);
    }

    static Node* zigzag(Node* g, Node* p, Node* x) {
        if (!g || !p || !x) return nullptr;
        g->left = zag(p, x);
        return zig(g, x);
    }

    static Node* zagzig(Node* g, Node* p, Node* x) {
        if (!g || !p || !x) return nullptr;
        g->right = zig(p, x);
        return zag(g, x);
    }

    static Node* insert(Node* root, int val) {
        if (!root) return nullptr;
        root->insert(val);
        root = find(root, val);
        return root;
    }

    static std::vector<Node*> split(Node* root, int k) {
        root = find(root, k);
        if (!root) return {};
        return {root->left, root->right};
    }

    static Node* join(Node* left, Node* right) {
        if (!left) return right;
        Node* max = left;
        while (max->right) {
            max = max->right;
        }
        left = find(left, max->val);
        if (!left) return nullptr;
        left->right = right;
        return left;
    }

    static Node* delete_node(Node* root, int val) {
        if (!root) return nullptr;
        root = find(root, val);
        if (!root) return nullptr;
        root = join(root->left, root->right);
        return root;
    }

    static Node* find(Node* root, int val, int i = 0) {
        i++;
        if (!root) return nullptr;
        Movement d1 = Movement::LEFT;
        Movement d2 = Movement::LEFT;
        if (root->val == val) {
            return root;
        }
        Node* g = root;
        Node* p;
        if (g->val < val) {
            p = g->right;
            d1 = Movement::RIGHT;
        } else {
            p = g->left;
            d1 = Movement::LEFT;
        }
        if (!p) return nullptr;
        if (p->val == val) {
            if (d1 == Movement::LEFT) {
                g = zig(g, p);
                return g;
            } else {
                g = zag(g, p);
                return g;
            }
        }
        Node* x;
        if (p->val < val) {
            x = p->right;
            d2 = Movement::RIGHT;
        } else {
            x = p->left;
            d2 = Movement::LEFT;
        }

        if (!x) return nullptr;

        if (p->val != val && x->val != val) {
            x = find(x, val, i);
            if (!x) return nullptr;
        }

        if (x->val == val) {
            if (d1 == Movement::LEFT) {
                if (d2 == Movement::LEFT) {
                    g = zigzig(g, p, x);
                    return g;
                } else {
                    g = zigzag(g, p, x);
                    return g;
                }
            } else {
                if (d2 == Movement::LEFT) {
                    g = zagzig(g, p, x);
                    return g;
                } else {
                    g = zagzag(g, p, x);
                    return g;
                }
            }
        }
        return nullptr;
    }
};

int main() {
    Node* tree = new Node(22);
    std::vector<int> vals = {10, 30, 6, 18, 27, 9, 15, 19, 12, 14};
    for (int val : vals) {
        tree = Splay::insert(tree, val);
    }

    tree->inorder();
    std::cout << std::endl;

    // tree = Splay::delete_node(tree, 22); // To delete
    tree = Splay::find(tree, 22); // To bring it to top
    // tree = Splay::insert(tree, 69); // To Insert and bring it to top

    if (tree) tree->inorder();
    std::cout << std::endl;

    if (tree) std::cout << tree->val << std::endl;

    delete tree;

    return 0;
}
